docs/proposals/20240807-in-place-updates.md

```diff
@@ -0,0 +1,488 @@
+    loop For all updaters in plan
+        mach->>hook: Run updater
+    end
+    mach->>apiserver: Mark Machine as updated
+end
+```
+
+The KCP external updates will work in a very similar way to MachineDeployments but removing the MachineSet level of indirection. In this case, it's the KCP controller the one in charge of both adding the update plan to the Machine and also updating the Machine spec. This follows this same pattern as for rolling updates, where the KCP controller directly creates and deletes Machines. Machines will be updates one by one, sequentially.
+
+
+### Machine updates
+
+```mermaid
+sequenceDiagram
+box Management Cluster
+    participant apiserver as kube-api server
+    participant capi as CAPI
+    participant mach as Machine Controller
+    participant hook as External updater 1
+    participant hook2 as External updater 2
+end
+
+box Workload Cluster
+    participant infra as Infrastructure
+end
+
+capi->>apiserver: Define Update Plan<br>and new Machine Spec
+apiserver->>mach: Notify changes
+mach->>hook: Start update
+hook->>infra: Update components
+loop Until finished
+    mach->>hook: finished?
+    hook->>mach: try in X secs
+end
+mach->>apiserver: Remove Updater 1 from plan
+mach->>hook2: Start update
+hook2->>infra: Update components
+loop Until finished
+    mach->>hook2: finished?
+    hook2->>mach: try in X secs
+end
+mach->>apiserver: Remove Updater 2 from plan
+mach->>apiserver: Mark Machine as up to date
+
+```
+
+Once a the Machine update plan is set and the Machine's spec has been updated with the desired changes, the Machine controller takes over. This controller is responsible for executing the right updaters in the right order, tracking the progress of those updaters and exposing this progress in the Machine status.
+
+The Machine controller will follow the order in which the names of the updaters have been defined in the Machine object, executing them one by one, sequentially.
+
+The controller will trigger updaters by hitting another RuntimeHook endpoint (eg. `/UpdateMachine`). This endpoint will take the Machine name as input. The updater could respond saying "update completed", "update failed" or "update in progress" with an optional "retry after X seconds". The CAPI controller will continuously poll the status of the update by hitting the same endpoint until it reaches a terminal state.
+
+CAPI expects the `/UpdateMachine` endpoint of an updater to be idempotent: for the same Machine with the same spec, the endpoint can be called any number of times (before and after it completes), and the end result should be the same. CAPI guarantees that once an `/UpdateMachine` endpoint has been called once, it won't change the Machine spec until the update reaches a terminal state.
+
+Once the update completes, the Machine controller will remove the name of the updater that has finished from the list of updaters and will start the next one. If the update fails, this will be reflected in the Machine status.
```

Comment:
It sounds like we're tracking state and keeping this state in the Machine controller itself. This is usually a common source of issues given that the state can drift from reality. Have we considered the set of hooks be only ever present on the MachineDeployment object, and the Machine object only contain its status, hence every updater has to be 1) re-entrant and 2) track where it "left-off".

This way, the status can be calculated from scratch at every iteration, rather than rely on sync calls and other means of strict operations.

Reply:
I don't think I follow. What state are you referring to? The list of updaters to be run?

Answering your other question, yeah we opted to have the set of hooks at the Machine level because that allows to reuse the same mechanism for both KCP and MD machines.

Regarding re-entrance for updaters: yeah, that is the idea here (it might need more clarification in the doc). CAPI will continue call the `/UpdateMachine` endpoint of an updater until this either returns success or failure. It's up to the updater to track the "update progress". Or maybe I didn't understand your comment correctly?

Reply:
From my understanding reading the proposal, it sounds like we're building a plan and tracking it in the Machine spec it self, which can be error prone; I'd suggest instead to find an approach that's ultimately declarative: declare the plan somewhere else and reflect the status of that plan in Machine status

Reply:
Ah I understand now, thanks for the clarification. Yeah this sounds very reasonable, let me give it a thought and I'll come back to it.